include/envoy/thread/thread.h

```diff
@@ -51,6 +51,27 @@ class ThreadFactory {
   virtual ThreadIdPtr currentThreadId() PURE;
 };
 
+/**
+ * A static singleton to the ThreadFactory corresponding to the build platform.
+ *
+ * The singleton must be initialized via set() early in main() with the appropriate ThreadFactory
+ * (see source/exe/{posix,win32}/platform_impl.h).
+ *
+ * This static singleton is an exception to Envoy's established practice for handling of singletons,
+ * which are typically registered with and accessed via the Envoy::Singleton::Manager. Reasons for
+ * the exception include drastic simplification of thread safety assertions; e.g.:
+ *   ASSERT(ThreadFactorySingleton::get()->currentThreadId() == original_thread_id_);
+ */
+class ThreadFactorySingleton {
+public:
+  static ThreadFactory* get() { return thread_factory_; }
+
+  static void set(ThreadFactory* thread_factory) { thread_factory_ = thread_factory; }
```

Comment:
Can you add an ASSERT here that thread_factory_ is nullptr? (I see you set to nullptr on exit from main, but unless we add null guards everywhere we use this it seems not worth it?)

Reply:
Done.